probe-rs/src/architecture/arm/ap/memory_ap/amba_ahb3.rs

```diff
@@ -0,0 +1,191 @@
+use crate::architecture::arm::{
+    ap::{AccessPortType, ApAccess, ApRegAccess},
+    communication_interface::RegisterParseError,
+    ArmError, DapAccess, FullyQualifiedApAddress, Register,
+};
+
+use super::{AddressIncrement, DataSize};
+
+/// Memory AP
+///
+/// The memory AP can be used to access a memory-mapped
+/// set of debug resources of the attached system.
+#[derive(Debug)]
+pub struct AmbaAhb3 {
+    address: FullyQualifiedApAddress,
+    csw: CSW,
+    cfg: super::registers::CFG,
+}
+
+impl AmbaAhb3 {
+    /// Creates a new AmbaAhb3 with `address` as base address.
+    pub fn new<P: DapAccess>(
+        probe: &mut P,
+        address: FullyQualifiedApAddress,
+    ) -> Result<Self, ArmError> {
+        use crate::architecture::arm::Register;
+        let csw = probe.read_raw_ap_register(&address, CSW::ADDRESS)?;
+        let cfg = probe.read_raw_ap_register(&address, super::registers::CFG::ADDRESS)?;
+        let (csw, cfg) = (csw.try_into()?, cfg.try_into()?);
+
+        let me = Self { address, csw, cfg };
+        let csw = CSW {
+            AddrInc: AddressIncrement::Single,
+            ..me.csw
+        };
+        probe.write_ap_register(&me, csw)?;
+        Ok(Self { csw, ..me })
+    }
+}
+
+impl super::MemoryApType for AmbaAhb3 {
+    type CSW = CSW;
+
+    fn status<P: ApAccess + ?Sized>(&mut self, probe: &mut P) -> Result<Self::CSW, ArmError> {
+        assert_eq!(super::registers::CSW::ADDRESS, CSW::ADDRESS);
```

Comment:
Why are there two definitions of CSW, how can we make this less confusing, and how can we turn this assert into a compile-time check?

Reply:
Each MemAP has its own CSW. I changed that to static_assertions.
Ideally I’d wish this could be done in the trait directly but this is using a trick with consts and `Self::CSW` from the trait cannot be used in constants.

Reply:
Boo... I dislike the repetition, maybe we should define a macro for which you only have to write the implementation defined fields... Otherwise this will be very tedious to maintain.

Reply:
Well, I don't expect any new MemAp type to show up nor these to change anytime soon :)

Macros don't play well with rustfmt nor rust-analyzer so they're annoying in their own way too.

Do you mind if we keep this as it is for now and eventually factories part of this register's definition in a future iteration ?

Reply:
> Macros don't play well with rustfmt nor rust-analyzer so they're annoying in their own way too.

This is a macro, though, right now :D

Sure, let's revisit this later.